Docs improvement Subscriptions

sdobbelaere opened the pull request:
As referred to in issue #375 some improved guidance on getting subscriptions running including the missing pieces related to the test-server.

bellini666 commented on docs/guide/subscriptions.md:
```diff
@@ -1,6 +1,136 @@
+
+## Making Django compatible
+
+It's important to realise that Django doesnt support websockets out of the box.
```
```suggestion
It's important to realise that Django doesn't support websockets out of the box.
```

bellini666 commented on docs/guide/subscriptions.md:
```diff
@@ -1,6 +1,136 @@
+It's important to realise that Django doesnt support websockets out of the box.
+To resolve this, we can help the platform along a little.
+
+Edit your `MyProject.asgi.py` file and replace it with the following content.
+Ensure that you replace the relevant code with your setup.
```
I think it is probably interesting to point out that [channels](https://channels.readthedocs.io/en/latest/) can be used to add websockets capabilities to django, and that the user can follow its instructions on how to setup it (as it will probably be more updated and have more content than us).

After that we can provide a minimal example to get started

bellini666 commented on docs/guide/subscriptions.md:
```diff
@@ -1,6 +1,136 @@
+
+urlpatterns = [
+	...
+    path('graphql/', AsyncGraphQLView.as_view(
+        schema=schema,
+        graphiql=settings.DEBUG,
+        subscriptions_enabled=True
+        )
+    ),
```
nit to what `black` would do

```suggestion
    path(
        'graphql/',
        AsyncGraphQLView.as_view(
            schema=schema,
            graphiql=settings.DEBUG,
            subscriptions_enabled=True,
        ),
    ),
```

bellini666 commented on docs/guide/subscriptions.md:
```diff
@@ -1,6 +1,136 @@
+
+## Setup local testing
+
+The classic `./manage.py runserver` will not support subscriptions. However, Django has daphne support out of the box to ensure that we can actually use Daphne for the runserver command. [Source](https://docs.djangoproject.com/en/4.2/howto/deployment/asgi/daphne/)
```
Maybe point out that the default `runserver` implementation runs on WSGI mode, meaning the channels integration would not work with it.

Then mention that about `daphne`, that is an ASGI protocol server, provides a custom `runserver` that runs on ASGI.

Maybe also important to mention that `daphne` can be used to help with local development, even if the user is running another ASGI protocol server in production, like `uvicorn`, `hypercorn`, etc

bellini666 commented on docs/guide/subscriptions.md:
````diff
@@ -1,6 +1,136 @@
+]
+```
+
+and add your ASGI_APPLICATION setting in your settings.py
````
nit

```suggestion
and add your `ASGI_APPLICATION` setting in your settings.py
```

bellini666 commented on docs/guide/subscriptions.md:
````diff
@@ -1,6 +1,136 @@
+Go and edit your schema-file and add:
+
+```python
+from strawberry import type, subscription
+import asyncio
+
+@type
+class Subscription:
+    @subscription
````
nit to align with the pattern we use across other documentations

```suggestion
import asyncio
import strawberry

@strawberry.type
class Subscription:
    @strawberry.subscription
```

bellini666 reviewed (changes requested):
Very nice in general! :)

sdobbelaere commented:
- Added the proposed improvements to the docs
- Added a new router that is authentication enabled. (Original router doesn't provide the user object, thus not very batteries included as it needs fixing straight out of the box)
- Consolidated how to find the user-object through get_current_user(info)

bellini666 commented on strawberry_django/auth/queries.py:
```diff
@@ -1,10 +1,28 @@
+def get_current_user(info):
+    """Get and return the current user based on various scenarios."""
+    try:
+        user = info.context.request.user
```
This can also cause errors when calling this inside an async function

Maybe also provide an `async def aget_current_user(info): return sync_to_async(get_current_user)(info)` bellow for async cases

bellini666 commented on strawberry_django/auth/queries.py:
```diff
@@ -1,10 +1,28 @@
+
 def resolve_current_user(info):
-    if not info.context.request.user.is_authenticated:
+    user = get_current_user()
```
```suggestion
    user = get_current_user(info)
```

bellini666 replied:
You marked this as resolved, but it is still missing the info =P

bellini666 commented on strawberry_django/auth/queries.py:
```diff
@@ -1,10 +1,28 @@
 import strawberry_django
 
 
+def get_current_user(info):
```
Also, it would be nice to add typing to the `info` argument and type the return value as `AbstractBaseUser`

sdobbelaere replied:
Importing AbstractBaseUser seems to be giving me some issues.  I'll skip on that for now. But I like the utils.py suggestion a lot, the typing I've added as well.

bellini666 commented on tests/projects/schema.py:
```diff
@@ -379,7 +380,7 @@ class Query:
     @strawberry_django.field
     def me(self, info: Info) -> Optional[UserType]:
-        user = info.context.request.user
+        user = get_current_user()
```
```suggestion
        user = get_current_user(info)
```

sdobbelaere commented on strawberry_django/routers.py:
```diff
@@ -0,0 +1,68 @@
+from __future__ import annotations
```
I looked at the original too, and that's the one that put me on the wrong trail related to authentication.  
As I was trying to make things more 'battery included' it seems like a wise idea to include the new version.  

Since the original doesn't leave much room for subclassing I decided to copy-paste it, and add the authentication to avoid breaking other peoples installs.

sdobbelaere commented on strawberry_django/auth/queries.py:
```diff
@@ -1,10 +1,28 @@
+            # When using this through websockets / subscriptions, scope sits inside of the request
+            user = info.context.get("request").scope.get("user")
+
+    return user
```
Sure, I'll do that.  Any thoughts on the id for it? Literally added XXX for now.

bellini666 commented on strawberry_django/auth/utils.py:
```diff
@@ -0,0 +1,23 @@
+            user = info.context.get("request").scope.get("user")
+
+    # Access an attribute inside the user object to force loading it in async contexts.
+    user.is_authenticated
```
You will probably need to make `ruff` happy here by adding a `noqa` to a rule

Try to run pre-commit locally as there are some issues to be solved: https://results.pre-commit.ci/run/github/308695886/1696080990.Uxt5tuPWSO2Hu7uePKlx4g

bellini666 replied:
Also, `user` might be `None` in here
```suggestion
    if user is not None:
        user.is_authenticated
```

bellini666 commented on strawberry_django/auth/utils.py:
```diff
@@ -0,0 +1,23 @@
+from strawberry.types import Info
+
+
+def get_current_user(info: Info):
```
Still missing the return annotation in here

I think it should be `Optional[AbstractUser]`

sdobbelaere replied:
importing `from django.contrib.auth.models import AbstractUser` or from django.contrib.auth.models import AbstractBaseUser won't work as the django application isn't fully loaded at this point.  

If you have a solution., happy to hear it.

bellini666 replied:
import it inside a `if TYPE_CHECKING:`

sdobbelaere replied:
Ruff keeps kicking out TYPE_CHECKING.  

I added:

```python
from typing import Optional, TYPE_CHECKING

from asgiref.sync import sync_to_async
from strawberry.types import Info

if TYPE_CHECKING:
    from django.contrib.auth.models import AbstractBaseUser
```

after committing, it turns into:

```python
from typing import Optional

from asgiref.sync import sync_to_async
from django.contrib.auth.models import AbstractBaseUser
from strawberry.types import Info
```

Thoughts?

bellini666 replied:
You probably need to annotate it as a string, like use `"AbstractBaseUser"` instead of `AbstractBaseUser`

That way ruff will allow you to leave it inside the `if TYPE_CHECKING`.

sdobbelaere replied:
Actually, I tried to run it before committing as well - and it just doesn't work. Django will not allow you to load the user-model outside of your Django-project.  Or at the very least before all apps are loaded, which causes similar issues.  

Perhaps you want to give it a shot yourself?  I've put as much time in this particular detail as I can.

bellini666 replied:
The solution was doing something like this:

![image](https://github.com/strawberry-graphql/strawberry-graphql-django/assets/134025/20ee7661-3478-46eb-855d-37493efceb39)

But having said that, I fixed it for you and actually used our `UserType`, which is a union of `AbstractBaseUser` and `AnonymousUser`

bellini666 commented on strawberry_django/auth/utils.py:
```diff
@@ -0,0 +1,23 @@
+
+
+async def aget_current_user(info: Info):
+    return sync_to_async(get_current_user)(info)
```
This needs to be imported

bellini666 commented on docs/guide/types.md:
````diff
@@ -103,12 +103,15 @@ You can use that `info` parameter to, for example,
 limit access to results based on the current user in the request:
 
 ```{.python title=types.py}
+from stawberry_django.auth.queries import get_current_user
````
This needs to be updated to the new path

```suggestion
from stawberry_django.auth.utils import get_current_user
```

bellini666 reviewed (changes requested):
Almost there :)

bellini666 commented:
@sdobbelaere I just solved the remaining pyright issues for you.

If the tests are passing fine, is there anything else missing from this or can it be merged?

bellini666 commented on docs/guide/subscriptions.md:
````diff
@@ -1,6 +1,140 @@
+
+```
+
+Note, django-channels allows for a lot more complexity. Here we just cover the basic framework to get
````
obs. pre-commit alex won't allow the word `just` to be used. It might need to be reworded here.

https://results.pre-commit.ci/run/github/308695886/1696869627.klCewzb2Qp6uMthM-uq5iw

sdobbelaere replied:
Thanks for the help.  I've adjusted the wording.  
I did add a get_current_user override to allow for quick testing earlier today.  If you like what I did, let me know and I'll add it to the docs as well.

sdobbelaere commented:
> @sdobbelaere I just solved the remaining pyright issues for you.
> 
> If the tests are passing fine, is there anything else missing from this or can it be merged?

There is commit b0a464a455afe3ca3f6fd27048d0e12a5c5c7a08 which allows you to manually set a user in the context for testing, on the condition that you mark it as "is_test_mode". 

Why did I add this?  Running tests needs access to a user in my case, this allows to construct a simple context when running tests like so:

```python
from django.contrib.auth import get_user_model
from MyProject.schemas import schema

user = get_user_model().objects.last()
dummy_context = {'is_test_mode': True, 'user': user}

resp = schema.execute_sync(query=query, context_value=dummy_context)
```

 Happy to add it to the docs.

bellini666 commented on strawberry_django/auth/utils.py:
```diff
@@ -0,0 +1,65 @@
+            except AttributeError:
+                # When running tests, we must manually add a context.
+                # For now let's grab it here:
+                is_test_mode = info.context.get("is_test_mode")
+                if is_test_mode:
```
I'm not a huge fan of this. Usually we would want to mock this or solve it in another way

What issue were you trying to fix with this code?

bellini666 replied:
Oh, I just saw your comment in the PR. I'll reply there

sdobbelaere replied:
I like the current approach much better indeed.  My addition was reverted, and the docs which were missing the `with` statement are fixed for whoever needs it after me.

bellini666 commented:
> > @sdobbelaere I just solved the remaining pyright issues for you.
> > If the tests are passing fine, is there anything else missing from this or can it be merged?
> 
> Thank you for the help, much appreciated. The only thing left is commit [b0a464a](https://github.com/strawberry-graphql/strawberry-graphql-django/commit/b0a464a455afe3ca3f6fd27048d0e12a5c5c7a08) which allows you to manually set a user in the context for testing, on the condition that you mark it as "is_test_mode".
> 
> Why did I add this? Running tests needs access to a user in my case, this allows to construct a simple context when running tests like so:
> 
> ```python
> from django.contrib.auth import get_user_model
> from MyProject.schemas import schema
> 
> user = get_user_model().objects.last()
> dummy_context = {'is_test_mode': True, 'user': user}
> 
> resp = schema.execute_sync(query=query, context_value=dummy_context)
> ```
> 
> Happy to add it to the docs.

As I mentioned there, I don't like that approach very much, specially because we have a way for forcing a user into the request: https://github.com/strawberry-graphql/strawberry-graphql-django/blob/main/strawberry_django/test/client.py#L75

You can see it being used in here for example: https://github.com/strawberry-graphql/strawberry-graphql-django/blob/main/tests/test_permissions.py#L49

bellini666 reviewed (changes requested):
Just a small nitpick and it should be good to merge :)

Thanks for taking the time to adjust everything I asked
